services/consensusalgo/benchmarkconsensus/algorithm.go

```diff
@@ -46,3 +55,83 @@ func (s *service) generateNewProposedBlock() (*protocol.BlockPairContainer, erro
 	}
 	return nil, nil
 }
+
+func (s *service) nonLeaderHandleCommit(blockPair *protocol.BlockPairContainer) {
+	err := s.nonLeaderValidateBlock(blockPair)
+	if err != nil {
+		s.reporting.Error(err) // TODO: wrap with added context
+		return
+	}
+	err = s.nonLeaderCommitAndReply(blockPair)
+	if err != nil {
+		s.reporting.Error(err) // TODO: wrap with added context
+		return
+	}
+}
+
+func (s *service) nonLeaderValidateBlock(blockPair *protocol.BlockPairContainer) error {
+	// nils
+	if blockPair.TransactionsBlock == nil ||
+		blockPair.ResultsBlock == nil ||
+		blockPair.TransactionsBlock.Header == nil ||
+		blockPair.ResultsBlock.Header == nil ||
+		blockPair.ResultsBlock.BlockProof == nil {
+		panic("invalid block: missing fields")
+	}
+
+	// block height
+	blockHeight := blockPair.TransactionsBlock.Header.BlockHeight()
+	if blockHeight > s.lastCommittedBlockHeight()+1 {
+		return errors.Errorf("invalid block: future block height %d", blockHeight)
+	}
+
+	// prev block hash ptr
+	if s.lastCommittedBlock != nil && blockHeight == s.lastCommittedBlockHeight()+1 {
+		prevTxHash := crypto.CalcTransactionsBlockHash(s.lastCommittedBlock)
+		if !blockPair.TransactionsBlock.Header.PrevBlockHashPtr().Equal(prevTxHash) {
+			return errors.Errorf("transactions prev block hash does not match prev block: %s", prevTxHash)
+		}
+		prevRxHash := crypto.CalcResultsBlockHash(s.lastCommittedBlock)
+		if !blockPair.ResultsBlock.Header.PrevBlockHashPtr().Equal(prevRxHash) {
+			return errors.Errorf("results prev block hash does not match prev block: %s", prevRxHash)
+		}
+	}
+
+	// block proof
+	if !blockPair.ResultsBlock.BlockProof.IsTypeBenchmarkConsensus() {
+		return errors.Errorf("incorrect block proof type: %s", blockPair.ResultsBlock.BlockProof.Type())
```

Comment:
Logically this check is suppose to be before the prev block hash ptr

Reply:
done